### DISCUSSION TITLE
how to get the data from useQuery in onSuccess callback?

### DISCUSSION THREAD

**Discussion by rookie-luochao** [TanStack/query#7953] [Ideas]
```
const { refetch } = useQuery({
    queryKey: ['todos'],
    queryFn: fetchTodos,
    onSuccess: (res) => { console.log(res) },
    onError: (error) => {
      if (error.status === 400) {
        toast.error('Something went wrong');
      } else {
        toast.error('Something else went wrong');
      }
    }
  });
```

**Comment by TkDodo** [comment]
callbacks on `useQuery` are deprecated in v4 and have been removed in v5:

- https://github.com/TanStack/query/discussions/5279

**Reply by rookie-luochao** [reply]
I think you can remove the onSuccess callback function, but you should provide a replacement hook, because onSuccess callback can provide good flexibility in some scenarios

**Reply by rookie-luochao** [reply]
> callbacks on `useQuery` are deprecated in v4 and have been removed in v5:
> 
> * [RFC: remove callbacks from useQuery #5279](https://github.com/TanStack/query/discussions/5279)

I raised this question because I found out through searching that onSuccess calllback has been removed

**Reply by TkDodo** [reply]
have you read the article? There is a specific example for error toasts, why the callbacks are bad for those, and what to do instead ...

**Reply by rookie-luochao** [reply]
> have you read the article? There is a specific example for error toasts, why the callbacks are bad for those, and what to do instead ...

ok, I'll read it carefully first to see if I can find what I think is an elegant solution, thanks for you

**Comment by rookie-luochao** [comment]
How can I elegantly cover scenarios that require onSuccess callback flexibility?

**Reply by TkDodo** [reply]
you'll need to be more specific with that question because there is no one-size-fits-all answer

**Reply by rookie-luochao** [reply]
> you'll need to be more specific with that question because there is no one-size-fits-all answer

For example, I am facing a scenario where I don’t need to get the data of useQuery immediately by default. I need to meet certain conditions to manually trigger an event to start a rotation. Then I need to call useQuery through rotation to get the data, and then process the data. The only way I can think of is to use useEffect with deps[data], which feels very strange and is far less concise than onSuccess callback.

**Reply by TkDodo** [reply]
so, lazy queries?

https://tanstack.com/query/latest/docs/framework/react/guides/disabling-queries#lazy-queries
